Softeq.XToolkit.PushNotifications.Droid/DroidPushNotificationsService.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Developed by Softeq Development Corporation
+// http://www.softeq.com
+
+using System;
+using System.Threading.Tasks;
+using Android.App;
+using Android.Content;
+using Firebase;
+using Firebase.Iid;
+using Firebase.Messaging;
+using Softeq.XToolkit.Common.Interfaces;
+
+namespace Softeq.XToolkit.PushNotifications.Droid
+{
+    public class DroidPushNotificationsService : PushNotificationsServiceBase, IDisposable
+    {
+        private readonly INotificationsSettingsProvider _notificationsSettings;
+        private readonly Context _appContext;
+
+        private bool _isInitialized;
+        private bool _showForegroundNotificationsInSystem;
+
+        private bool _registrationRequired;
+
+        public DroidPushNotificationsService(IRemotePushNotificationsService remotePushNotificationsService,
+            IPushTokenStorageService pushTokenStorageService, IPushNotificationsHandler pushNotificationsHandler,
+            IPushNotificationParser pushNotificationParser, INotificationsSettingsProvider notificationsSettings, ILogManager logManager)
+            : base(remotePushNotificationsService, pushTokenStorageService, pushNotificationsHandler, pushNotificationParser, logManager)
+        {
+            _appContext = Application.Context;
+            _notificationsSettings = notificationsSettings;
+        }
+
+        public override void Initialize(bool showForegroundNotificationsInSystem)
+        {
+            if (_isInitialized)
+            {
+                Logger.Debug("PushNotificationsServiceDroid: Already Initialized");
+                return;
+            }
+
+            _isInitialized = true;
+            _showForegroundNotificationsInSystem = showForegroundNotificationsInSystem;
+
+            //TODO: + update on locale changed
+            NotificationsHelper.CreateNotificationChannels(_appContext, _notificationsSettings);
+
+            FirebaseApp.InitializeApp(_appContext);
+            XFirebaseIIDService.OnTokenRefreshed += OnPushTokenRefreshed;
+            XFirebaseMessagingService.OnNotificationReceived += OnNotificationReceived;
+        }
+
+        public override void RegisterForPushNotifications()
+        {
+            var token = FirebaseInstanceId.Instance.Token;
+            if (token == null)
+            {
+                _registrationRequired = true;
+            }
+            else
+            {
+                _registrationRequired = false;
+                OnRegisteredForPushNotificaions(token);
+            }
+        }
+
+        protected override void UnregisterFromPushTokenInSystem()
+        {
+            // TODO: possibly use topics and UnsubscribeFromTopic instead
+            Task.Run(() =>
+            {
+                try
+                {
+                    FirebaseInstanceId.Instance.DeleteInstanceId(); //Throws exception if there's no Internet Connection
```

Comment:
should we handle only Internet exception or all exceptions?

Reply:
replaced Exception with IOException which is actually thrown by Firebase when there is no Internet